homeassistant/components/sonos/sensor.py

```diff
@@ -224,3 +246,89 @@ async def _async_update_state(self) -> None:
             "items": {fav.item_id: fav.title for fav in self.favorites}
         }
         self.async_write_ha_state()
+
+
+class SonosNextAlarmEntity(SonosEntity, SensorEntity):
+    """Representation of a Sonos next alarm entity."""
+
+    _attr_device_class = SensorDeviceClass.TIMESTAMP
+    _attr_has_entity_name = True
+    _attr_translation_key = "next_alarm"
+
+    def __init__(self, speaker: SonosSpeaker, config_entry: SonosConfigEntry) -> None:
+        """Initialize the next alarm sensor."""
+        super().__init__(speaker, config_entry)
+        self._attr_unique_id = f"{self.soco.uid}-next-alarm"
+        self.household_id = speaker.household_id
+
+    async def async_added_to_hass(self) -> None:
+        """Handle sensor setup when added to hass."""
+        await super().async_added_to_hass()
+        self.async_on_remove(
+            async_dispatcher_connect(
+                self.hass,
+                f"{SONOS_ALARMS_UPDATED}-{self.household_id}",
+                self.async_write_ha_state,
+            )
+        )
+
+        async def async_write_state_daily(now: datetime.datetime) -> None:
+            """Update state attributes each calendar day."""
+            _LOGGER.debug("Daily update of next alarm for %s", self.name)
+            self.async_write_ha_state()
+
+        self.async_on_remove(
+            async_track_time_change(
+                self.hass, async_write_state_daily, hour=0, minute=0, second=0
+            )
+        )
+
+    async def _async_fallback_poll(self) -> None:
+        """Call the central alarm polling method."""
+        alarms: SonosAlarms = self.config_entry.runtime_data.alarms[self.household_id]
+        assert alarms.async_poll
+        await alarms.async_poll()
+
+    @property
+    def icon(self) -> str:
+        """Icon of the entity."""
+        if self.native_value is None:
+            return "mdi:alarm-off"
+        if self._following_alarm_trigger is None:
+            return "mdi:alarm"
+        return "mdi:alarm-multiple"
+
+    @property
+    def native_value(self) -> datetime.datetime | None:
+        """Return the next alarm time."""
+        return self.speaker.alarms.get_next_alarm_datetime(zone_uid=self.soco.uid)
+
+    @property
+    def _is_today(self) -> bool:
+        """Return whether the next alarm is scheduled for today."""
+        if self.native_value is None:
+            return False
+        return self.native_value.date() == dt_util.now().date()
+
+    @property
+    def _following_alarm_trigger(self) -> datetime.datetime | None:
+        """Return the following alarm after the current next one."""
+        if self.native_value is None:
```

Comment:
We use self.native_value alot - and it calls into the Soco function and loops through the alarm list.  On each update we will call that function at least 4 times. Can we cache it for performance?